drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	if (vring >= (int)priv->hw_nr_virtqs) {
+		DRV_LOG(ERR, "Too big vring id: %d.", vring);
+		return -E2BIG;
+	}
+
+	//TO_DO: check if vid set here is sutiable
+	priv->vid = vid;
+
+	if(virtio_pci_dev_get_status(priv->vpdev)&VIRTIO_CONFIG_STATUS_DRIVER_OK) {
+		DRV_LOG(ERR, "Can not set vring stat when driver ok vDPA device: %s.", vdev->device->name);
+		return -EINVAL;
+	}
+
+	pthread_mutex_lock(&priv->vq_config_lock);
+	if (!state && priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_disable(priv, vring);
+	}
+	if (state && !priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_enable(priv, vring);
+	}
+	if (state && priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_disable(priv, vring);
+		if(ret) {
+			DRV_LOG(ERR, "Fail to set vring state to 0, ret:%d vring:%d state:%d.", ret, vring, state);
+			pthread_mutex_unlock(&priv->vq_config_lock);
+			return ret;
+		}
+		ret = virtio_vdpa_virtq_enable(priv, vring);
+	}
+	pthread_mutex_unlock(&priv->vq_config_lock);
+	if(ret) {
+		DRV_LOG(ERR, "Fail to set vring state, ret:%d vring:%d state:%d.", ret, vring, state);
+		return ret;
+	}
+
+	DRV_LOG(INFO, "vDPA device %d  set vring %d state %d.", vid, vring, state);
+	return 0;
+}
+
+static int
+virtio_vdpa_dma_unmap(struct virtio_vdpa_priv *priv)
+{
+	uint32_t i;
+	int ret;
+	struct rte_vhost_memory *mem = NULL;
+	int vfio_container_fd;
+
+	ret = rte_vhost_get_mem_table(priv->vid, &mem);
+	if (ret < 0) {
+		DRV_LOG(ERR, "Failed to get VM memory layout.");
+		goto exit;
+	}
+
+	vfio_container_fd = priv->vfio_container_fd;
+
+	for (i = 0; i < mem->nregions; i++) {
+		struct rte_vhost_mem_region *reg;
+
+		reg = &mem->regions[i];
+		DRV_LOG(INFO, "%s, region %u: HVA 0x%" PRIx64 ", "
+			"GPA 0x%" PRIx64 ", size 0x%" PRIx64 ".",
+			"DMA unmap", i,
+			reg->host_user_addr, reg->guest_phys_addr, reg->size);
+
+		ret = rte_vfio_container_dma_unmap(vfio_container_fd,
+			reg->host_user_addr, reg->guest_phys_addr,
+			reg->size);
+		if (ret < 0) {
+			DRV_LOG(ERR, "DMA unmap failed.");
+			goto exit;
+		}
+	}
+
+exit:
+	free(mem);
+	return ret;
+}
+
+static int
+virtio_vdpa_dma_map(struct virtio_vdpa_priv *priv)
+{
+	uint32_t i;
+	int ret;
+	struct rte_vhost_memory *mem = NULL;
+	int vfio_container_fd;
+
+	ret = rte_vhost_get_mem_table(priv->vid, &mem);
+	if (ret < 0) {
+		DRV_LOG(ERR, "Failed to get VM memory layout.");
+		goto exit;
+	}
+
+	vfio_container_fd = priv->vfio_container_fd;
+
+	for (i = 0; i < mem->nregions; i++) {
+		struct rte_vhost_mem_region *reg;
+
+		reg = &mem->regions[i];
+		DRV_LOG(INFO, "%s, region %u: HVA 0x%" PRIx64 ", "
+			"GPA 0x%" PRIx64 ", size 0x%" PRIx64 ".",
+			"DMA map", i,
+			reg->host_user_addr, reg->guest_phys_addr, reg->size);
+
+		ret = rte_vfio_container_dma_map(vfio_container_fd,
+			reg->host_user_addr, reg->guest_phys_addr,
+			reg->size);
+		if (ret < 0) {
+			DRV_LOG(ERR, "DMA map failed.");
+			goto exit;
```

Comment:
you have partially mapped and partially unmap.

add error unwinding code to do dma_unmap on when it fails half way.